### PR TITLE
feat: support generate_subscripts function

### LINKED ISSUE #11575: support function `generate_subscripts`
https://www.postgresql.org/docs/current/functions-srf.html

```
generate_subscripts ( array anyarray, dim integer ) → setof integer
Generates a series comprising the valid subscripts of the dim'th dimension of the given array.
```

```
generate_subscripts ( array anyarray, dim integer, reverse boolean ) → setof integer
Generates a series comprising the valid subscripts of the dim'th dimension of the given array. When reverse is true, returns the series in reverse order.
```

### PR DESCRIPTION
**Pull request by lmatz**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Resolve #11575

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR contains user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/expr/src/table_function/generate_subscripts.rs [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+#[function("generate_subscripts(list, int32, boolean) -> setof int32")]
+fn generate_subscripts_reverse(
+    array: ListRef<'_>,
+    dim: i32,
+    reverse: bool,
+) -> Result<impl Iterator<Item = i32>> {
+    let (mut cur, end) = generate_subscripts_inner(array, dim, reverse);
+
+    // if we don't branch reverse inside, then we have to box it outside to have the same type
+    // different closures different anonymous types, or range and reverse range.
```
Comment (xiangjinwu) [inline src/expr/src/table_function/generate_subscripts.rs:69]: FYI there is https://docs.rs/auto_enums/0.8.2/auto_enums/

File: src/expr/src/table_function/generate_subscripts.rs [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    match dim {
+        ..=0 => nothing,
+        1 => {
+            if reverse {
+                (array.len() as i32, 0)
+            } else {
+                (1, array.len() as i32 + 1)
+            }
+        }
+        // Although RW's array can be zig-zag, we just look at the first element.
```
Comment (xiangjinwu) [inline src/expr/src/table_function/generate_subscripts.rs:140]: There are several alternatives:
* length of first element (this PR)
* max length of all elements
* length of any element when rectangular, or error when not
* always error on dim >= 2 (DuckDB and our current `array_length`)

We may keep the current behavior as experimental/undocumented, and make a breaking change of the zig-zag behavior later.

**Review by xiangjinwu — APPROVED** [review]
rest lgtm

File: src/expr/src/table_function/generate_subscripts.rs (hunk 1 above)
Comment (lmatz) [reply src/expr/src/table_function/generate_subscripts.rs]: thanks, I use `auto_enum` instead.

File: src/expr/src/table_function/generate_subscripts.rs (hunk 2 above)
Comment (lmatz) [reply src/expr/src/table_function/generate_subscripts.rs:134]: thanks! let me create an issue to track this after the PR gets merged

**Pull request merged**